parachain/pallets/outbound-queue/src/lib.rs

```diff
@@ -92,153 +111,184 @@ pub mod pallet {
+			xcm_hash: XcmHash,
+		},
+		MessageAccepted {
+			/// ID of the XCM message
+			xcm_hash: XcmHash,
+			/// The nonce assigned to this message
+			nonce: u64,
+		},
+		MessagesCommitted {
+			/// Merkle root of the committed messages
+			root: H256,
+			/// number of committed messages
+			count: u64,
+		},
 	}
 
 	#[pallet::error]
 	pub enum Error<T> {
-		/// The message payload exceeds byte limit.
-		PayloadTooLarge,
-		/// No more messages can be queued for the channel during this commit cycle.
-		QueueSizeLimitReached,
-		/// Cannot increment nonce
-		Overflow,
+		/// The message is too large
+		MessageTooLarge,
 	}
 
-	/// Interval between commitments
+	/// Messages to be committed every block. This storage value is killed at the start of every
+	/// block, so should never go into block PoV.
+	///
+	/// Is never read in the runtime, only by offchain code.
+	///
+	/// Inspired by the `frame_system::Pallet::Events` storage value
 	#[pallet::storage]
-	#[pallet::getter(fn interval)]
-	pub(super) type Interval<T: Config> = StorageValue<_, T::BlockNumber, ValueQuery>;
+	#[pallet::unbounded]
+	pub(super) type Messages<T: Config> = StorageValue<_, Vec<Box<Message>>, ValueQuery>;
 
-	/// Messages waiting to be committed.
 	#[pallet::storage]
-	pub(super) type MessageQueue<T: Config> = StorageValue<
-		_,
-		BoundedVec<Message<T::MaxMessagePayloadSize>, T::MaxMessagesPerCommit>,
-		ValueQuery,
-	>;
+	#[pallet::unbounded]
+	pub(super) type MessageLeaves<T: Config> = StorageValue<_, Vec<H256>, ValueQuery>;
 
 	#[pallet::storage]
 	pub type Nonce<T: Config> = StorageMap<_, Twox64Concat, ParaId, u64, ValueQuery>;
 
-	#[pallet::genesis_config]
-	pub struct GenesisConfig<T: Config> {
-		pub interval: T::BlockNumber,
-	}
-
-	#[cfg(feature = "std")]
-	impl<T: Config> Default for GenesisConfig<T> {
-		fn default() -> Self {
-			Self { interval: Default::default() }
-		}
-	}
-
-	#[pallet::genesis_build]
-	impl<T: Config> GenesisBuild<T> for GenesisConfig<T> {
-		fn build(&self) {
-			<Interval<T>>::put(self.interval);
-		}
-	}
-
 	#[pallet::hooks]
 	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T>
 	where
 		T::AccountId: AsRef<[u8]>,
 	{
-		// Generate a message commitment when the chain is idle with enough remaining weight
-		// The commitment hash is included in an [`AuxiliaryDigestItem`] in the block header,
-		// with the corresponding commitment is persisted offchain.
-		fn on_idle(_n: T::BlockNumber, total_weight: Weight) -> Weight {
-			let weight_remaining = total_weight.saturating_sub(T::WeightInfo::on_commit(
-				T::MaxMessagesPerCommit::get(),
-				T::MaxMessagePayloadSize::get(),
-			));
-			if weight_remaining.ref_time() <= MINIMUM_WEIGHT_REMAIN_IN_BLOCK.ref_time() {
-				return total_weight
-			}
-			Self::commit(total_weight)
+		fn on_initialize(_: T::BlockNumber) -> Weight {
+			// Remove storage from previous block
+			Messages::<T>::kill();
+			MessageLeaves::<T>::kill();
+			return T::WeightInfo::on_finalize()
```

Comment:
The way it works is that if a pallet has an `on_finalize` handler, it needs to return the weight of `on_finalize` in `on_initialize`. This reserves some block weight for `on_finalize` to run successfully.

For that reason, you'll notice that the `on_finalize` does not return the weight it consumed, since it's scheduled to run right at the end of block production.

Reply:
then should it be `T::WeightInfo::on_finalize()` + 2 write in `on_initialize` ?

Reply:
Not sure if deleting storage counts as a read, will check,